Feature/idp 823 explicit contract first flow (#21)
gsoft-inc/wl-openapi-msbuild: README.md

# Workleap.OpenApi.MSBuild

Validates at build time that the OpenAPI specification files extracted from the ASP.NET Core Web API being built conform to Workleap API guidelines.

> Depending if the users the Contract-First or Code-First approach this MsBuild task will:

Reviewer: ```suggestion
Depending if the user chose the Contract-First or Code-First development mode this MSBuild task will:
```

> - Install tools: OasDiff, Spectral, SwashbuckleCLI

Reviewer: Maybe provide links to the tools?

- https://github.com/Tufin/oasdiff
- https://github.com/stoplightio/spectral
- https://github.com/domaindrivendev/Swashbuckle.AspNetCore?tab=readme-ov-file#swashbuckleaspnetcorecli

- Generate the OpenAPI specification file from the associated Web API

> - Validate Spectral rules

Reviewer: Maybe add link? https://github.com/gsoft-inc/wl-api-guidelines/blob/main/.spectral.yaml

- Compare the given OpenAPI specification file with the generated one

> ## How it work

Reviewer: ```suggestion
## How it works
```

> - The entry point is `ValidateOpenApiTask.ExecuteAsync()` and will be executed during the build process on project that reference it. This is defined in `./src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets` as a `UsingTask.TaskName`

Reviewer: ```suggestion
- The entry point is `ValidateOpenApiTask.ExecuteAsync()` and will be executed after the referencing project is built. This is defined in `./src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets` as a `UsingTask.TaskName`
```

> - The entry point is `ValidateOpenApiTask.ExecuteAsync()` and will be executed during the build process on project that reference it. This is defined in `./src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets` as a `UsingTask.TaskName`

Reviewer: Maybe link to the official MSBuild docs? https://learn.microsoft.com/en-us/visualstudio/msbuild/tutorial-custom-task-code-generation?view=vs-2022#include-msbuild-properties-and-targets-in-a-package

- The default value are defined in the property group on the target `ValidateOpenApi` in this file `./src/Workleap.OpenApi.MSBuild/msbuild/tools/Workleap.OpenApi.MSBuild.targets`

## How to test locally

### How debug the project

Since it's a MSBuild task named `ValidateOpenApi` you can run it this command: `msbuild /t:ValidateOpenApi`. `./src/WebApiDebugger` already have a `launchSettings.json` named `ValidateOpenApi` that you can run in debug.

### With the system test

This command `./Run-SystemTest.ps1` will:

1. Pack the MSBuild library

> 2. Build the WebApi.MssBuild.SystemTest and inject the previously packed library

Reviewer: ```suggestion
2. Build the WebApi.MsBuild.SystemTest and inject the previously packed library
```

Sadly we used lowercase 's' in MSBuild...
